javascript/documentation/library-customization.rst

```diff
@@ -0,0 +1,249 @@
+Customizing the JavaScript analysis
+===================================
+
+This document describes the main extension points offered by the JavaScript analysis for customizing
+analysis behavior without editing the queries or libraries themselves.
+
+Customization mechanisms
+------------------------
+
+The two mechanisms used for customization are subclassing and overriding.
+
+By subclassing an abstract class used by the JavaScript analysis and implementing its abstract
+member predicates we can teach the analysis to handle further instances of abstract concepts it
+already understands. For example, the standard library defines an abstract class
+``SystemCommandExecution`` that covers various APIs for executing operating-system commands. This
+class is used by the command-injection analysis to identify problematic flows where input from a
+potentially malicious user is interpreted as the name of a system command to execute. By defining
+additional subclasses of ``SystemCommandExecution``, we can make this analysis more powerful without
+touching its implementation.
+
+By overriding a member predicate defined in the library, we can change its behavior either for all
+its receivers or only a subset. For example, the standard library predicate
+``ControlFlowNode::getASuccessor`` implements the basic control-flow graph on which many further
+analyses are based. By overriding it, we can add, suppress or modify control-flow graph edges.
+
+Once a customization has been defined, it needs to be brought into scope so that the default
+analysis queries pick it up. This can be done by adding the customizing definitions to
+``Customizations.qll``, an initially empty library file that is imported by the default library
```

Comment:
Should we explain how one in practice would implement these customizations for a query pack that can not modify `Customizations.qll`?

Reply:
I don't actually know how one would do that.

Reply:
Create `CompanyCustomizations.qll` and import that in all of the query pack queries is the way we do it at the moment. This obviously does not improve the builtin queries unless they are copied to the query pack.

Reply:
Hm, OK, that seems a little outside the scope of this document.